File: framework/src/datamirai_engine/tools/builtin/filesystem/glob_files.py
"""filesystem/glob — find files by glob pattern."""

from __future__ import annotations

import glob as _glob
import os
from typing import Any

from datamirai_engine.core.context import ExecutionContext
from datamirai_engine.tools.base import (
    BaseTool,
    ConfigField,
    ToolInput,
    ToolOutput,
    ToolSpec,
)


class GlobFilesTool(BaseTool):
    spec = ToolSpec(
        tool_type="filesystem/glob",
        version="1.0.0",
        display_name="Glob (Find Files)",
        description="Finds files matching a glob pattern (e.g. '**/*.py', 'src/**/*.ts'). Returns paths sorted by modification time (newest first).",
        category="filesystem",
        icon="folder-search",
        intents=[
            "buscar archivos por patron de nombre",
            "encontrar todos los archivos de un tipo",
            "listar archivos que coincidan con un glob",
        ],
        inputs=[
            ToolInput(name="pattern", type="string", required=True, description="Glob pattern (e.g. '**/*.py')"),
        ],
        outputs=[
            ToolOutput(name="files", type="array", description="List of matching file paths"),
            ToolOutput(name="count", type="number", description="Number of matches"),
        ],
        config=[
            ConfigField(name="path", type="string", default=".", description="Base directory to search in"),
            ConfigField(name="max_results", type="number", default=200, description="Maximum number of results"),
        ],
    )

    async def execute(
        self,
        inputs: dict[str, Any],
        config: dict[str, Any],
        context: ExecutionContext | None,
    ) -> dict[str, Any]:
        pattern = inputs["pattern"]
        base = os.path.abspath(config.get("path", "."))
        max_results = int(config.get("max_results", 200))

        full_pattern = os.path.join(base, pattern)
        matches = _glob.glob(full_pattern, recursive=True)

        # Filter to files only, sort by mtime descending
        files = [f for f in matches if os.path.isfile(f)]
        files.sort(key=lambda f: os.path.getmtime(f), reverse=True)
        files = files[:max_results]

        return {
            "files": files,
            "count": len(files),
        }

File: engine/src/tools/builtin/filesystem/glob_files.rs
//! GlobFilesTool

use super::*;

// ===========================================================================
// GlobFilesTool
// ===========================================================================

fs_tool! {
    struct GlobFilesTool, factory GlobFilesFactory;
    tool_type = "filesystem/glob_files",
    name = "Glob (Find Files)",
    description = "Finds files matching a glob pattern. Returns paths sorted by modification time (newest first).",
    category = "filesystem",
    inputs = [
        field("pattern", FieldType::String, true, "Glob pattern (e.g. '**/*.py')"),
    ],
    outputs = [
        field("files", FieldType::Array, true, "List of matching file paths"),
        field("count", FieldType::Number, true, "Number of matches"),
    ],
    config_fields = [
        field("path", FieldType::String, false, "Base directory to search in"),
        field("max_results", FieldType::Number, false, "Maximum number of results"),
    ]
}

#[async_trait]
impl Tool for GlobFilesTool {
    async fn execute(
        &self,
        inputs: HashMap<String, Value>,
        config: &HashMap<String, Value>,
        _context: &dyn ExecutionContext,
    ) -> Result<HashMap<String, Value>, ToolError> {
        let pattern = inputs
            .get("pattern")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::ExecutionFailed {
                tool_type: "filesystem/glob_files".into(),
                message: "missing required input: pattern".into(),
            })?;

        let base = config.get("path").and_then(|v| v.as_str()).unwrap_or(".");
        let max_results = config
            .get("max_results")
            .and_then(|v| v.as_u64())
            .unwrap_or(200) as usize;

        let base_path = fs::canonicalize(base).map_err(|e| ToolError::ExecutionFailed {
            tool_type: "filesystem/glob_files".into(),
            message: format!("Base path not found: {base} ({e})"),
        })?;

        let full_pattern = format!("{}/{}", base_path.display(), pattern);

        let paths = glob::glob(&full_pattern).map_err(|e| ToolError::ExecutionFailed {
            tool_type: "filesystem/glob_files".into(),
            message: format!("Invalid glob pattern: {e}"),
        })?;

        // Collect files with mtime
        let mut files_with_mtime: Vec<(PathBuf, f64)> = Vec::new();
        for entry in paths {
            let p = match entry {
                Ok(p) => p,
                Err(_) => continue,
            };
            if !p.is_file() {
                continue;
            }
            let mtime = p
                .metadata()
                .ok()
                .and_then(|m| m.modified().ok())
                .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
                .map(|d| d.as_secs_f64())
                .unwrap_or(0.0);
            files_with_mtime.push((p, mtime));
        }

        // Sort by mtime descending (newest first)
        files_with_mtime.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
        files_with_mtime.truncate(max_results);

        let files: Vec<String> = files_with_mtime
            .iter()
            .map(|(p, _)| p.display().to_string())
            .collect();
        let count = files.len();

        let mut out = HashMap::new();
        out.insert("files".to_string(), json!(files));
        out.insert("count".to_string(), json!(count));
        Ok(out)
    }
}
